{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c5d9343c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All tests completed successfully\n",
      "Question1: 3\n",
      "Question2: 2\n",
      "Question3: \n",
      "904746\n",
      "Question4: \n",
      "7161\n"
     ]
    }
   ],
   "source": [
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "\n",
    "def editDistance(x, y):\n",
    "    \"\"\"Returns the edit distance between two strings, x and y\"\"\"\n",
    "    # Create distance matrix\n",
    "    D = []\n",
    "    for i in range(len(x)+1):\n",
    "        D.append([0]*(len(y)+1))\n",
    "    \n",
    "    # Initialize first row and column of matrix\n",
    "    for i in range(len(x)+1):\n",
    "        D[i][0] = i\n",
    "    for i in range(len(y)+1):\n",
    "        D[0][i] = i\n",
    "    \n",
    "    # Fill in the rest of the matrix\n",
    "    for i in range(1, len(x)+1):\n",
    "        for j in range(1, len(y)+1):\n",
    "            distHor = D[i][j-1] + 1\n",
    "            distVer = D[i-1][j] + 1\n",
    "            if x[i-1] == y[j-1]:\n",
    "                distDiag = D[i-1][j-1]\n",
    "            else:\n",
    "                distDiag = D[i-1][j-1] + 1\n",
    "            D[i][j] = min(distHor, distVer, distDiag)\n",
    "    # Edit distance is the value in the bottom right corner of the matrix\n",
    "    return D[-1][-1]\n",
    "\n",
    "def bestApproximateMatchEditDistance(p, t):\n",
    "    \"\"\"Returns the edit distance between two strings, p and t\"\"\"\n",
    "    # Create distance matrix\n",
    "    D = []\n",
    "    for i in range(len(p)+1):\n",
    "        D.append([0]*(len(t)+1))\n",
    "    \n",
    "    # Initialize first row and column of matrix\n",
    "    for i in range(len(p)+1):\n",
    "        D[i][0] = i\n",
    "    # See slide 4 on  0440_approx__editdist3.pdf\n",
    "    # First row is already initialised to zero so we simply just comment the following two lines.\n",
    "    #for i in range(len(p)+1):\n",
    "    #    D[0][i] = i\n",
    "    \n",
    "    # Fill in the rest of the matrix\n",
    "    for i in range(1, len(p)+1):\n",
    "        for j in range(1, len(t)+1):\n",
    "            distHor = D[i][j-1] + 1\n",
    "            distVer = D[i-1][j] + 1\n",
    "            if p[i-1] == t[j-1]:\n",
    "                distDiag = D[i-1][j-1]\n",
    "            else:\n",
    "                distDiag = D[i-1][j-1] + 1\n",
    "            D[i][j] = min(distHor, distVer, distDiag)\n",
    "\n",
    "    # Best Approximate Match Distance is the smallest value of the last row\n",
    "    return min(D[-1])\n",
    "\n",
    "def test1():\n",
    "    \"\"\"P = GCGTATGC within T = TATTGGCTATACGGTT had 2 edits.\"\"\"\n",
    "    assert bestApproximateMatchEditDistance('GCGTATGC', 'TATTGGCTATACGGTT') == 2\n",
    "\n",
    "def question1(t):\n",
    "    \"\"\"What is the edit distance of the best match between pattern GCTGATCGATCGTACG and the excerpt of human chromosome 1? (Don't consider reverse complements.)\"\"\"\n",
    "    print(\"Question1: \" + str(bestApproximateMatchEditDistance('GCTGATCGATCGTACG', t)))\n",
    "\n",
    "def question2(t):\n",
    "    \"\"\"What is the edit distance of the best match between pattern GATTTACCAGATTGAG and the excerpt of human chromosome 1? (Don't consider reverse complements.)\"\"\"\n",
    "    print(\"Question2: \" + str(bestApproximateMatchEditDistance('GATTTACCAGATTGAG', t)))\n",
    "\n",
    "\n",
    "\n",
    "def overlap(a, b, min_length=3):\n",
    "    \"\"\" Return length of longest suffix of 'a' matching a prefix of 'b' that is at least 'min_length' characters long.  If no such overlap exists, return 0. \"\"\"\n",
    "    start = 0  # start all the way at the left\n",
    "    while True:\n",
    "        start = a.find(b[:min_length], start)  # look for b's prefix in a\n",
    "        if start == -1:  # no more occurrences to right\n",
    "            return 0\n",
    "        # found occurrence; check for full suffix/prefix match\n",
    "        if b.startswith(a[start:]):\n",
    "            return len(a)-start\n",
    "        start += 1  # move just past previous match\n",
    "\n",
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  # skip name line\n",
    "            seq = fh.readline().rstrip()  # read base sequence\n",
    "            fh.readline()  # skip placeholder line\n",
    "            qual = fh.readline().rstrip() # base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n",
    "\n",
    "def overlap_all_pairs(reads, min_length):\n",
    "    from pprint import pprint\n",
    "    \n",
    "    overlap_map = {}\n",
    "    overlap_graph = {}\n",
    "    overlap_pairs = []\n",
    "\n",
    "    #We use a Python dictionary to associate each k-mer with its corresponding set. \n",
    "    suffixDict = {}\n",
    "    for read in reads:\n",
    "        kmers = getkmers(read, min_length)\n",
    "        #print(kmers)\n",
    "        #(1) For every k-mer in a read, we add the read to the set object corresponding to that k-mer.\n",
    "        for kmer in kmers:\n",
    "            if not kmer in suffixDict.keys():\n",
    "                #Let every k-mer in the dataset have an associated Python set object, which starts out empty. \n",
    "                suffixDict[kmer] = set()\n",
    "            suffixDict[kmer].add(read)\n",
    "    #pprint(suffixDict)\n",
    "\n",
    "    #(2) Now, for each read a, we find all overlaps involving a suffix of a\n",
    "    for read in reads:\n",
    "        #we take a's length-k suffix,\n",
    "        suffix = read[-min_length:]\n",
    "        #if len(suffix) < min_length:\n",
    "        #    continue\n",
    "\n",
    "        #find all reads containing that k-mer (obtained from the corresponding set) ...\n",
    "        matching_reads = suffixDict[suffix]\n",
    "\n",
    "        #...and call overlap(a, b, min_length=k) for each.\n",
    "        for read2 in matching_reads:\n",
    "            # The most important point is that we do not call overlap(a, b, min_length=k) if b does not contain the length-k suffix of a.\n",
    "            #if read2.find(suffix) >= 0 and read2 != suffix :\n",
    "            if read2 != read :\n",
    "                val = overlap(read, read2, min_length) \n",
    "                if val > 0:\n",
    "                    overlap_map[ (read, read2) ] = val\n",
    "                    overlap_graph[read] = read2\n",
    "                    overlap_pairs.append( (read,read2) )\n",
    "\n",
    "    #pprint(overlap_map)\n",
    "    #pprint(overlap_pairs)\n",
    "    #pprint(overlap_graph)\n",
    "    return overlap_pairs, overlap_map, overlap_graph\n",
    "\n",
    "\n",
    "\n",
    "def getkmers(read, kmer_length):\n",
    "    \"\"\"our read is GATTA and k=3, we would add GATTA to the set objects for GAT, ATT and TTA.\"\"\"\n",
    "    return [ read[i:i+kmer_length] for i in range(len(read)+1-kmer_length) ]\n",
    "\n",
    "def test2():\n",
    "    kmers = getkmers('GATTA', 3)\n",
    "    assert kmers == ['GAT', 'ATT', 'TTA']\n",
    "\n",
    "def example1():\n",
    "    reads = ['ABCDEFG', 'EFGHIJ', 'HIJABC']\n",
    "    assert overlap_all_pairs(reads, 3)[0] == [('ABCDEFG', 'EFGHIJ'), ('EFGHIJ', 'HIJABC'), ('HIJABC', 'ABCDEFG')]\n",
    "    assert overlap_all_pairs(reads, 4)[0] == []\n",
    "\n",
    "def example2():\n",
    "    from pprint import pprint\n",
    "\n",
    "    reads = ['CGTACG', 'TACGTA', 'GTACGT', 'ACGTAC', 'GTACGA', 'TACGAT']\n",
    "\n",
    "    results4, overlap_map4, overlap_graph4 = overlap_all_pairs(reads, 4)\n",
    "    expected4 = [('CGTACG', 'TACGTA'),\n",
    "                 ('CGTACG', 'GTACGT'),\n",
    "                 ('CGTACG', 'GTACGA'),\n",
    "                 ('CGTACG', 'TACGAT'),\n",
    "                 ('TACGTA', 'ACGTAC'),\n",
    "                 ('TACGTA', 'CGTACG'),\n",
    "                 ('GTACGT', 'TACGTA'),\n",
    "                 ('GTACGT', 'ACGTAC'),\n",
    "                 ('ACGTAC', 'GTACGA'),\n",
    "                 ('ACGTAC', 'GTACGT'),\n",
    "                 ('ACGTAC', 'CGTACG'),\n",
    "                 ('GTACGA', 'TACGAT')]\n",
    "\n",
    "    assert sorted(results4) ==  sorted(expected4) , \"example2, first assert failed\"\n",
    "\n",
    "    results5, overlap_map5, overlap_graph5 = overlap_all_pairs(reads, 5)\n",
    "    expected5 = [('CGTACG', 'GTACGT'),\n",
    "                 ('CGTACG', 'GTACGA'),\n",
    "                 ('TACGTA', 'ACGTAC'),\n",
    "                 ('GTACGT', 'TACGTA'),\n",
    "                 ('ACGTAC', 'CGTACG'),\n",
    "                 ('GTACGA', 'TACGAT')] \n",
    "\n",
    "    assert sorted(results5) ==  sorted(expected5), \"example2, second assert failed\"\n",
    "\n",
    "def question3and4(reads):\n",
    "    from pprint import pprint\n",
    "\n",
    "    overlap_pairs, overlap_map, overlap_graph = overlap_all_pairs(reads, 30)\n",
    "\n",
    "    \"\"\"Picture the overlap graph corresponding to the overlaps just calculated. How many edges are in the graph? \n",
    "    In other words, how many distinct pairs of reads overlap?\"\"\"\n",
    "    print('Question3: ')\n",
    "    print(len(overlap_map))\n",
    "\n",
    "    \"\"\"Picture the overlap graph corresponding to the overlaps computed for the previous question. \n",
    "    How many nodes in this graph have at least one outgoing edge? (In other words, how many reads have a suffix involved in an overlap?)\"\"\"\n",
    "    print('Question4: ')\n",
    "    print(len(overlap_graph))\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "def main():\n",
    "    test1()\n",
    "    print(\"All tests completed successfully\")\n",
    "    genome = readGenome('chr1.GRCh38.excerpt.fasta')\n",
    "    question1(genome)\n",
    "    question2(genome)\n",
    "\n",
    "    test2()\n",
    "    example1()\n",
    "    example2()\n",
    "    reads, qualities = readFastq('ERR266411_1.for_asm.fastq')\n",
    "    question3and4(reads) \n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe49e870",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
